File: C++/PermsOfArr.cpp
/* Generate all permutations of a distinct int arr
Input: int[] Output: int[] perms Constraints: 1<= arr length <= 6
Backtracking method*/
#include <iostream>
#include <vector>
using namespace std;

void Backtrack(vector<int>& nums, vector<int>& used, vector<int>& path, vector<vector<int>>& res){
    if(path.size() == nums.size()){
        res.push_back(path);
        return;
    }

    for(int i = 0; i < nums.size(); i++){
        if(used[i] == 1) continue;

        used[i] = 1;
        path.push_back(nums[i]);
        Backtrack(nums, used, path, res);
        path.pop_back();
        used[i] = 0;
    }
}

vector<vector<int>> Permute(vector<int> nums){
    vector<vector<int>> res;
    vector<int> used(nums.size(), 0);
    vector<int> path;
    Backtrack(nums, used, path, res);
    return res;
}

int main() {
    vector<int> nums = {1, 2, 3};
    auto Perms = Permute(nums);

    cout << "Permutations:" << endl;
    for (auto& p : Perms) {
        for (int v : p) cout << v << " ";
        cout << endl;
    }
    // Expected count: 6
    return 0;
}

File: Python/PermsOfArr.py
# Generate all permutations of a distinct int arr
# Input: int[] Output: int[] perms Constraints: 1<= arr length <= 6
# Backtracking method

def Permute(nums):
    res = []
    used = [0] * len(nums)

    def Backtrack(path):
        if len(path) == len(nums):
            res.append(path.copy())
            return
        
        for i in range(len(nums)):
            if used[i] == 1:
                continue

            used[i] = 1
            path.append(nums[i])
            Backtrack(path)
            path.pop()
            used[i] = 0

    Backtrack([])
    return res

def main():
    nums = [1, 2, 3]
    perms = Permute(nums)

    print("Permutations:")
    for p in perms:
        print(p)
    # Expected count: 6

if __name__ == "__main__":
    main()